pep-0657.rst

```diff
@@ -0,0 +1,404 @@
+
+    Traceback (most recent call last):
+      File "test.py", line 2, in <module>
+        x['a']['b']['c']['d'] = 1
+        ^^^^^^^^^^^^^^^^
+    TypeError: 'NoneType' object is not subscriptable
+
+indicating to the user that the object ``x['a']['b']`` must have been ``None``.
+This highlighting will occur for every frame in the traceback. For instance, if
+a similar error is part of a complex function call chain, the traceback would
+display the code associated to the current instruction in every frame::
+
+    Traceback (most recent call last):
+      File "test.py", line 14, in <module>
+        lel3(x)
+        ^^^^^^^
+      File "test.py", line 12, in lel3
+        return lel2(x) / 23
+               ^^^^^^^
+      File "test.py", line 9, in lel2
+        return 25 + lel(x) + lel(x)
+                    ^^^^^^
+      File "test.py", line 6, in lel
+        return 1 + foo(a,b,c=x['z']['x']['y']['z']['y'], d=e)
+                             ^^^^^^^^^^^^^^^^^^^^^
+    TypeError: 'NoneType' object is not subscriptable
+
+This problem presents itself quite a lot when passing down multiple objects to
+function calls while accessing the same attribute in them. For instance, this
+error::
+
+    Traceback (most recent call last):
+      File "test.py", line 19, in <module>
+        foo(a.name, b.name, c.name)
+    AttributeError: 'NoneType' object has no attribute 'name'
+
+With the improvements in this PEP this would show::
+
+    Traceback (most recent call last):
+      File "test.py", line 17, in <module>
+        foo(a.name, b.name, c.name)
+                    ^^^^^^
+    AttributeError: 'NoneType' object has no attribute 'name'
+
+
+These sorts of ambiguous error messages are commonplace when dealing with lines
+with complex mathematical expressions, especially with libraries such as numpy
+where arithmetic operations can fail based on the arguments. For example: ::
+
+    Traceback (most recent call last):
+      File "test.py", line 1, in <module>
+        x = (a + b) @ (c + d)
+      ValueError: operands could not be broadcast together with shapes (1,2) (2,3)
+
+There is no clear indication as to which operation failed, was it the addition
+on the left, the right or the matrix multiplication in the middle? With this
+PEP the new error message would look like::
+
+    Traceback (most recent call last):
+      File "test.py", line 1, in <module>
+        x = (a + b) @ (c + d)
+                       ^^^^^
+      ValueError: operands could not be broadcast together with shapes (1,2) (2,3)
+
+Giving a much clearer and easier to debug error message.
+
+
+Debugging aside, this extra information would also be useful for coverage code
+coverage tools, enabling them to measure expression-level coverage instead of
+just line-level coverage. For instance, given the following line: ::
+
+    x = foo() if bar() else baz()
+
+coverage, profile or state analysis tools will highlight the full line in both
```

Comment:
```suggestion
coverage, profile or static analysis tools will highlight the full line in both
```
Although I'm not sure what tools this would be referring to. Static analysis tools can operate on the AST which already has column offset information.

Reply:
No, this is *state* analysis tools: tools that analyzer the state of a given program, like getting the current frame stack and the such.

Reply:
I half suspected that's what you meant :) It's a pretty obscure term in that case; I googled it and mostly got stuff some business development process.

Reply:
Maybe we can toss in an "(e.g debuggers)" to help clarify